File: 2022/day4.py

from aoc import input_as_lines, parse_positive_ints_str

inp = input_as_lines("day4.txt")


# PART 1
ret = 0
for line in inp:
    pairs = parse_positive_ints_str(line)
    x0 = pairs[0]
    x1 = pairs[1]
    y0 = pairs[2]
    y1 = pairs[3]
    if (x0 <= y0 and x1 >= y1) or (y0 <= x0 and y1 >= x1):
        ret += 1
print(ret)

# PART 2
ret = 0
for line in inp:
    pairs = parse_positive_ints_str(line)
    x0 = pairs[0]
    x1 = pairs[1]
    y0 = pairs[2]
    y1 = pairs[3]
    one = set(range(x0, x1 + 1))
    two = set(range(y0, y1 + 1))
    if len(one & two) != 0:
        ret += 1
print(ret)

File: 2022/day4.dart
import 'aoc.dart';

void main() {
  var input = inputAsLines(4, test: false);
  int ret = 0;

  // Part 1
  for (String line in input) {
    var pairs = parsePositiveIntsFromString(line);
    int x0 = pairs[0];
    int x1 = pairs[1];
    int y0 = pairs[2];
    int y1 = pairs[3];
    if ((x0 <= y0 && x1 >= y1) || (y0 <= x0 && y1 >= x1)) {
      ret += 1;
    }
  }
  print(ret);

  ret = 0;

  // Part 2
  for (String line in input) {
    var pairs = parsePositiveIntsFromString(line);
    int x0 = pairs[0];
    int x1 = pairs[1];
    int y0 = pairs[2];
    int y1 = pairs[3];
    var one = [for (var i = x0; i <= x1; i++) i].toSet();
    var two = [for (var i = y0; i <= y1; i++) i].toSet();
    if (one.intersection(two).length != 0) {
      ret += 1;
    }
  }
  print(ret);
}
